modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/ExpressionUtilsImpl.java

```diff
@@ -93,56 +93,138 @@ public Object evaluate(final Node question, final Map<String, Object> values, fi
     private ExpressionUtilsImpl.ParsedExpression parseExpressionInputs(final String expression,
         final Map<String, Object> values)
     {
-        String expr = expression;
-
-        Boolean missingValue = false;
-        Map<String, ExpressionArgument> questions = new HashMap<>();
-
-        int start = expr.indexOf(START_MARKER);
-        int end = expr.indexOf(END_MARKER, start);
-        // For each argument in the expression, parse the question name and default value if present.
-        // To prevent question names from breaking the evaluating funtion, replace them with a default
-        // argument name in the evaluated expression.
-        while (start > -1 && end > -1) {
-            int defaultStart = expr.indexOf(DEFAULT_MARKER, start);
-            boolean hasDefault = defaultStart > -1 && defaultStart < end;
-
-            // Parse out the question name and default value if provided
-            String questionName;
-            String defaultValue = null;
-            if (hasDefault) {
-                questionName = expr.substring(start + START_MARKER.length(), defaultStart);
-                defaultValue = expr.substring(defaultStart + DEFAULT_MARKER.length(), end);
-            } else {
-                questionName = expr.substring(start + START_MARKER.length(), end);
+        ExpressionParser parser = new ExpressionParser(expression, values);
+        return parser.parse();
+    }
+
+    private final class ExpressionParser
+    {
+        private String expression;
+        private final Map<String, Object> questionValues;
+        private boolean missingValue;
+        private Map<String, ExpressionArgument> questions = new HashMap<>();
+
+        // Next argument details
+        private boolean isArrayArgument;
+        private String startMarker;
+        private String endMarker;
+        private int start;
+        private int end;
+
+
+        ExpressionParser(String expression, final Map<String, Object> questionValues)
+        {
+            this.expression = expression;
+            this.questionValues = questionValues;
+
+            scanNextArgument();
+        }
+
+        private void scanNextArgument()
+        {
+            this.isArrayArgument = isNextArgumentMultivalued(this.expression);
+            this.startMarker = this.isArrayArgument ? START_MARKER_ARRAY : START_MARKER_SINGLE;
+            this.endMarker = this.isArrayArgument ? END_MARKER_ARRAY : END_MARKER_SINGLE;
+            this.start = this.expression.indexOf(this.startMarker);
+            this.end = this.expression.indexOf(this.endMarker);
```

Comment:
It is possible to have a `}` somewhere in the expression before the start marker, since that is a very common syntax element.

```suggestion
            this.end = this.expression.indexOf(this.endMarker, this.start);
```